### PR TITLE
Fix handling of aggregate literals with only a right-hand comparison term (Issue #311)

### LINKED ISSUE #311: Lower bound cannot be given after the aggregate
Putting the lower bound of an aggregate _after_ an aggregate instead of _before_ it results in a `NullPointerException`.

For example, the following program:
```
p(1).
p(2).
q :- #count { N : p(N) } > 2.
```
results in:
```
Exception in thread "main" java.lang.NullPointerException
        at at.ac.tuwien.kr.alpha.grounder.transformation.aggregates.AggregateOperatorNormalization.rewriteAggregateOperator(AggregateOperatorNormalization.java:81)
        at at.ac.tuwien.kr.alpha.grounder.transformation.aggregates.AggregateOperatorNormalization.rewriteLiteral(AggregateOperatorNormalization.java:63)
        at at.ac.tuwien.kr.alpha.grounder.transformation.aggregates.AggregateOperatorNormalization.normalize(AggregateOperatorNormalization.java:56)
        at at.ac.tuwien.kr.alpha.grounder.transformation.aggregates.AggregateRewriting.apply(AggregateRewriting.java:77)
        at at.ac.tuwien.kr.alpha.grounder.transformation.NormalizeProgramTransformation.apply(NormalizeProgramTransformation.java:31)
        at at.ac.tuwien.kr.alpha.api.Alpha.normalizeProgram(Alpha.java:125)
        at at.ac.tuwien.kr.alpha.Main.main(Main.java:98)
```

Related to #268.

### PR DESCRIPTION
**Pull request by madmike200590**
Fixed in `AggregateOperatorNormalization`:
Aggregates of form `#aggr{...} OP TERM` are rewritten into `TERM inv(OP) #aggr{...}` and then processed as before.
`inv(OP)` refers to the inverse operator, i.e. `inv(<) = >=` etc. For operators `=` and `!=` no inverse operator is used, i.e. rewritten form is `TERM OP #aggr{...}`.

Closes #311

### CODE REVIEW

File: alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/solver/AggregatesTest.java [hunk 1]
```diff
@@ -335,4 +335,36 @@ public void aggregateSumGlobalVariable(RegressionTestConfig cfg) {
 				"full(2)");
 	}
 
+	@AggregateRegressionTest
+	public void aggregateRightHandTermCountGreater(RegressionTestConfig cfg) {
+		String program = "p(1)." +
```
Comment (lorenzleutgeb) [inline alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/solver/AggregatesTest.java:340]: @AntoniusW regarding #301: Here, IMO [JEP-378: Text Blocks](https://openjdk.java.net/jeps/378) would make this look prettier.

File: alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/aggregates/AggregateOperatorNormalization.java [hunk 2]
```diff
@@ -69,6 +69,24 @@ private AggregateOperatorNormalization() {
...
+					atom.getUpperBoundTerm(), 
+					atom.getAggregateFunction(), 
+					atom.getAggregateElements()).toLiteral(!lit.isNegated());
+			} else {
+				flipped = Atoms.newAggregateAtom(
+					atom.getUpperBoundOperator().negate(), 
+					atom.getUpperBoundTerm(), 
+					atom.getAggregateFunction(), 
+					atom.getAggregateElements()).toLiteral(!lit.isNegated());
+			}
```
Comment (lorenzleutgeb) [inline alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/aggregates/AggregateOperatorNormalization.java:87]: The difference in the second argument seems to be the only difference between the two branches. You could make that more explicit by doing:

```suggestion
			var flippedUpperBoundOperator = EQ.equals(operator) || NE.equals(operator) ? atom.getUpperBoundOperator() : atom.getUpperBoundOperator().negate();
                        // "flip" right-hand operator and term to left-hand ones.
			AggregateLiteral flipped = Atoms.newAggregateAtom(
					flippedUpperBoundOperator, 
					atom.getUpperBoundTerm(), 
					atom.getAggregateFunction(),
					atom.getAggregateElements()).toLiteral(!lit.isNegated());
```

**Review by lorenzleutgeb — APPROVED** [review]
LGTM, minor remarks.

**Comment by AntoniusW** [comment]
> Fixed in `AggregateOperatorNormalization`: Aggregates of form `#aggr{...} OP TERM` are rewritten into `TERM inv(OP) #aggr{...}` and then processed as before. `inv(OP)` refers to the inverse operator, i.e. `inv(<) = >=` etc. For operators `=` and `!=` no inverse operator is used, i.e. rewritten form is `TERM OP #aggr{...}`.

Uhm, I beg to disagree:
`#agg {..} > 2` should be equal to `2 < #agg {..}`, i.e., what is needed here is flipping of operators, not negation.

Consider for example `X < Y` which is flipped `Y > X` and not `Y >= X`, to see that consider the case with `X=2,Y=2` and we have `2 < 2` is false as is `2 > 2` while `2 >= 2` becomes true. The unit tests currently do not consider such a case and hence cannot show that the implementation is bugged at the moment.

Please change that.

**Review by rtaupe — CHANGES_REQUESTED** [review]
Please address @AntoniusW's comment.

**Comment by madmike200590** [comment]
> > Fixed in `AggregateOperatorNormalization`: Aggregates of form `#aggr{...} OP TERM` are rewritten into `TERM inv(OP) #aggr{...}` and then processed as before. `inv(OP)` refers to the inverse operator, i.e. `inv(<) = >=` etc. For operators `=` and `!=` no inverse operator is used, i.e. rewritten form is `TERM OP #aggr{...}`.
> 
> Uhm, I beg to disagree: `#agg {..} > 2` should be equal to `2 < #agg {..}`, i.e., what is needed here is flipping of operators, not negation.
> 
> Consider for example `X < Y` which is flipped `Y > X` and not `Y >= X`, to see that consider the case with `X=2,Y=2` and we have `2 < 2` is false as is `2 > 2` while `2 >= 2` becomes true. The unit tests currently do not consider such a case and hence cannot show that the implementation is bugged at the moment.
> 
> Please change that.

Thanks for catching this! Since the actual rewriting in `AggregateOperatorNormalization` uses operator negation in quite a few places, I mentally tripped myself up here :sweat_smile: 

This should now be fixed.

**Review by rtaupe — APPROVED** [review]
Thanks for fixing this! 😃

**Pull request merged**